TechnicalDocuments/Proposals/Feedback Navigation.md

```diff
@@ -0,0 +1,81 @@
+# Feedback Navigation
+
+* Author(s): Sergey Shulga
+* Review Manager: TBD
+
+## Introduction
+
+As all of you probably know in our codebase we use `Route` based navigation, where VM has a signal of routes which emits a specific route and FlowController just listens to it and performs navigation.
+
+Although the current concept is fine and we've been happy with it for a while now, it does look a bit weird when it comes to integration with RAF.
+For example, we've recently agreed that we derive all routes from the `State`  of the VM. Also, sometimes we need to pass closures to the routes as we need to notify our VM when a user selects something on the child screen.
+
+```swift
+routes = state.signal.skipRepeats().filterMap { state in
+            switch state {
+            case let .loaded(_, error?):
+                return .showAlert(.make(error: error))
+            case .searching:
+                return .showMap(mapResponsesObserver.send(value:)) // here we are passing a closure which will push value into our VM
+            case let .showingDirections(_, place):
+                return .showDirections(place)
+            }
+        }
+```
+
+## Motivation
+
+The idea is to deal with navigation in the same way we deal with networking. Where navigation can be treated as an async operation the result of which is going to impact the current state of the VM.
+
+## Proposed solution
+
+```swift    
+static func whenSearching(flowController: PharmaciesFlowController) -> Feedback<State, Event> {
+        return Feedback { state -> SignalProducer<Event, NoError> in
```

Comment:
I think it will worth mentioning that this will allow a better handling on situations when some state transitions shouldn't trigger routes, for example going back in navigation state and transitioning from `state2` back to `state1` should not trigger route that is usually triggered when transitioning to `state1` from `state0`. We should be able to isolate this logic into specific feedbacks instead of doing that when `route` signal is defined. We can even add some helper methods in future to make it more convenient to declare. Same with cancelation of navigation side effect.

Btw this constructor should be using proposed `flatMapFirst` constructor, I guess.